php/libraries/Database.class.inc

```diff
@@ -500,16 +517,26 @@ class Database
         $err = $this->trackChanges($table, array(), $where, 'D');
 
         $this->_printQuery($query);
-        $this->affected = $this->_PDO->exec($query);
-        if ($this->affected === false) {
+        $result = $this->_PDO->exec($query);
+        if ($result === false) {
             throw new DatabaseException(
                 "Database DELETE did not execute successfully."
                 . print_r($this->_PDO->errorInfo(), true),
                 $query,
                 $where
             );
-
+        } elseif ($result === 0) {
+            if (DEBUG) {
+                fwrite(
+                    STDERR,
+                    "Database DELETE executed successfully "
+                    . "but 0 matched rows: " . $query. "\n"
+                );
+                return;
+            }
         }
+
+        $this->affected = $result;
```

Comment:
I wanted to discuss the "DEBUG" constant actually. A lot of my
```
        } elseif ($result === 0) {
             if (DEBUG) {
```
's won't actually get executed(current code flow-wise, not DB execution-wise). I was thinking of throwing an exception at first if 0 results were affected but that's a bit too dramatic in my opinion, so I left it as is for now. I think it is important to be aware of instances where nothing was matched, but not enough for it halt execution of the rest of the code. There are many ways to handle this such as throwing an exception, writing to the error logs, writing # results affected to the history table(which I don't think gets looked at very frequently...), printing additional info upon showdatabasequeries=1, etc, but I'm not sure what level of notice everyone thinks would be appropriate.